doc/man/bbctl.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH BBCTL 1 "January 24, 2018"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
bbctl \- query and control tool for BitBabbler hardware RNG devices
.SH SYNOPSIS
.B bbctl
.RI [ options ]


.SH DESCRIPTION
The \fBbbctl\fP program can be used to issue command requests to the control
socket of software controlling a BitBabbler device (such as the \fBseedd\fP(1)
daemon).


.SH OPTIONS
The following options are available:

.TP
.B \-s, \-\-scan
Scan for active devices.  This will report the device identifiers which can be
queried from the owner of the control socket.

.TP
.BI "\-i, \-\-device\-id=" id
Act on only the specified device.  If no devices are explicitly specified then
the default is to act upon all of them.  This option may be passed multiple
times to act on some subset of the available devices.  The \fIid\fP must be an
identifier name as reported by \fBbbctl\ \-\-scan\fP, you cannot use device
logical or physical addresses here.

.TP
.B \-b, \-\-bin\-freq
Report the 8-bit symbol frequencies.

.TP
.B \-B, \-\-bin\-freq16
Report the 16-bit symbol frequencies.

.TP
.B "    \-\-bin\-count"
Report the 8-bit symbol counts.  Similar to \fB\-\-bin\-freq\fP except the bins
are reported in symbol order instead of sorted by frequency.

.TP
.B "    \-\-bin\-count16"
Report the 16-bit symbol counts.  Similar to \fB\-\-bin\-freq16\fP except the
bins are reported in symbol order instead of sorted by frequency.

.TP
.BI "    \-\-first=" n
Show only the first \fIn\fP results.  Useful when you don't want to actually
see all 65 thousand entries for the 16-bit bins.  The default (if neither this
nor the \fB\-\-last\fP option are specified) is to report everything in its
full glory.  Don't say I didn't warn you.

.TP
.BI "    \-\-last=" n
Show only the last \fIn\fP results.  Useful when you don't want to actually
see all 65 thousand entries for the 16-bit bins.  If used together with the
\fB\-\-first\fP option, then both the requested head and tail of the results
will be shown.

.TP
.B \-r, \-\-bit\-runs
Report on runs of consecutive bits.

.TP
.B \-S, \-\-stats
Report general QA statistics.

.TP
.BI "\-c, \-\-control\-socket=" path
The filesystem path for the service control socket to query.  This can belong
to any process that supports the BitBabbler control socket interface and for
which the user running \fBbbctl\fP has permission to connect to.

An address of the form \fItcp:host:port\fP may be used if the control socket is
bound to a TCP port rather than a unix domain socket path.  The \fIhost\fP part
can be a DNS hostname or address literal.  If an IPv6 address literal is used
it should be enclosed in square brackets (e.g. tcp:[::1]:2020 to bind to port
2020 on the local IPv6 interface).  The \fIport\fP can be a port number or a
service name (as defined in \fI/etc/services\fP or other system name-service
databases which are queried by \fBgetaddrinfo\fP(3)).

.TP
.BI "\-V, \-\-log\-verbosity=" n
Change the logging verbosity of the control socket owner.

.TP
.BI "    \-\-waitfor=" device : passbytes : retry : timeout
This option will make \fBbbctl\fP wait before exiting until the \fBseedd\fP(1)
QA checking reports that at least \fIpassbytes\fP of good entropy have been
obtained from the given \fIdevice\fP.  It will check for that every \fIretry\fP
milliseconds, waiting for a maximum of \fItimeout\fP milliseconds before
failing.

The \fIdevice\fP is a QA test identifier as reported by \fB\-\-scan\fP, and
must be provided, as must the expected \fIpassbytes\fP count.  The \fIretry\fP
time is optional, and if not specified it will default to 1000 milliseconds.
If the \fItimeout\fP is 0 (or not explicitly passed), then this will wait for
an unbounded amount of time for the requested condition to occur.

The \fIpassbytes\fP, \fIretry\fP, and \fItimeout\fP parameters may be suffixed
with an SI multiplier (e.g. k, M, G) as a convenience, so a \fItimeout\fP of
30k would wait for 30 seconds.

This option may be passed multiple times to wait for multiple devices, and the
given conditions for each of them will be tested for in the order that they
are specified on the command line.  i.e. Later conditions will not be tested
for at all until all prior ones have been met, and the \fItimeout\fP clock for
each test only begins after the previous test has successfully completed.

When all required conditions pass, \fBbbctl\fP will report success with an exit
code of 0.  If a \fItimeout\fP is exceeded, or any other error occurs which
means the test cannot be successfully completed (like passing a \fIdevice\fP
which does not exist, or querying a \fB\-\-control\-socket\fP which no process
provides), then a non-zero exit code will be returned.

This option mostly exists to make it possible to delay or even prevent other
services from starting until a sufficient amount of entropy has been obtained
to feel comfortable that they can operate securely or as intended.  See the
notes on \fBBOOT\ SEQUENCING\fP in \fBseedd\fP(1) for more details on that.
It may be used for other purposes too, but note that \fIpassbytes\fP is an
absolute measure of the number of good bytes seen since \fBseedd\fP was
started, it is not relative to the number that were obtained prior to
executing this request.

.TP
.B \-v, \-\-verbose
Make more noise about what is going on internally.  It may be passed multiple
times to get swamped with even more information.

.TP
.B \-?, \-\-help
Show a shorter version of all of this, which may fit on a single page.

.TP
.B "    \-\-version"
Report the \fBbbctl\fP release version.


.SH FILES
.TP
.I /run/bit\-babbler/seedd.socket
The default \fB\-\-control\-socket\fP path if not explicitly specified.  This
may be under \fI/var/run\fP on platforms which don't (yet) provide a \fI/run\fP
top level directory (or a TCP socket on platforms which don't support unix
domain sockets).  It is set at compile time by \fBSEEDD_CONTROL_SOCKET\fP.


.SH SEE ALSO
.BR seedd (1).


.SH AUTHOR
.B seedd
was written by Ron <ron@debian.org>.  You can send bug reports, feature requests,
praise and complaints to support@bitbabbler.org.



debian/changelog
bit-babbler (0.9) unstable; urgency=medium

  * Refactor the udev rules for kernel 4.12 bind events and the mess udev made
    with handling introduction of a new event type.

  * Install udev rules and systemd units with make install, and fix the paths
    in them at configure time.  There should be no reason for that needing to
    be Debian specific anymore.

  * Ignore -Wgnu-designator at the sites where we use them rather than just
    disabling it globally.  It's good to mark Special Things as being special.

  * Drop the open coded set of 'hardening flags' from the Debian packaging,
    we test for and include those as part of the normal build system options.

  * Mark up exceptions for UBSan's not-actually-undefined tests so that we can
    apply them to catch any code where that behaviour might be unintentional.
    We do have deliberate use for Inf and NaN results in some statistics, and
    for 'implicit' modulo 2^32 in a uint32_t.  This makes those more explicit.

  * Be explicit about more traditionally implicit type conversions too so we
    can usefully employ warnings about unintended/unanalysed conversions that
    could be narrowing or reduce precision.

  * Use RefuseManualStart/RefuseManualStop in seedd-wait.service.  It is a(n
    optional) sequence point for seeding the kernel at boot time, so there is
    no reason or value in attempting to start or stop it after that. Stopping
    it could even be actively harmful, as that would also stop any unit that
    did Require it to complete before they started, even though what they
    actually Require (a freshly seeded kernel) has been satisfied and will
    not be changed.

  * Build-Depends: debhelper (>= 9), debhelper (>= 9.20160709) | dh-systemd
    We need this little abomination because dh-systemd has been merged with
    debhelper 9.20160709, but Jessie is still ELTS promised until 30/6/2022,
    and it only has debhelper 9.20150101+deb8u2 ...  So when move-fast-and-
    break-shit dropped dh-systemd from buster, it broke shit for people who
    want packages to build on at least all supported releases without silly
    avoidable busywork ...  Closes: #958581

  * Enable bbvirt to log to syslog, which is useful to see what happens when
    it's called from udev rules and the like which don't already log output
    sent to stderr.

  * Try to work around systemd deadlocking the boot with libvirt from Bullseye
    when it starts a zombie .socket unit before the service unit dependencies
    are satisfied - causing anything which attempts to communicate with it to
    just hang until the systemd time out "boot ordering bug solver" kicks in.
    We can only make the race a bit harder to lose here, a real fix involves
    declaring proper ordering dependencies for the libvirt socket unit (and
    fixing the networking.service unit to not abandon trying to bring the
    network up if the ifupdown-pre.service times out).

  * Chop the NUL terminator off control socket messages before passing them to
    JSON::XS for decoding, the 4.0 release now considers them trailing garbage
    rather than just ignoring them like earlier releases did.

 -- Ron Lee <ron@debian.org>  Wed, 27 Oct 2021 17:19:23 +1030

bit-babbler (0.8) unstable; urgency=medium

  * Support hotplugging devices into libvirt guest domains which have names
    containing characters that are not valid as part of a shell variable name.
    Another reminder that the important part of keeping things as simple as
    possible is always the "as possible" bit.

  * Support reading seedd(1) options from a configuration file.  The original
    design plan explicitly avoided this, partly just to keep the code as
    simple and easy to audit as possible, and partly because it was desirable
    to make invocation as simple and foolproof as possible.  The more options
    that something has, the easier it is to make some mistake with running it
    which could have subtle and even serious consequences.  But we are at the
    point now where there are enough real alternative options which are either
    genuinely desirable or needed for some use case, that the balance becomes
    weighted toward being able to keep persistent configuration settings in a
    file rather than having to spell them out on the command line each time.

    The final straw for making this change now was the inability of systemd to
    sanely support the existing simplified configuration interface that was
    provided in /etc/default/seedd for the SysV init script.  When given the
    alternative choices available to us of either adding a shell wrapper to
    do what systemd could not, or forcing people to manually edit or override
    the systemd unit directly to make any configuration change, this was
    clearly the Lesser Evil to embrace if we were going to provide a native
    systemd unit for the system daemon.  The former gains us nothing over the
    existing LSB init script, and the latter would require every user to first
    have a solid grasp of all the non-obvious consequences which can come into
    play when configuring a system which (according to systemd.directives(7))
    "contains 2464 entries in 13 sections, referring to 241 individual manual
    pages" - and where even package maintainers and systemd upstream still
    make mistakes that can take a long time for the real consequences to be
    noticed.  So if we were to provide a systemd unit, it needs to be well
    tested and give people few, if any, reasons to ever need to modify it.

  * Preserve existing configuration on package upgrades.  The new default
    configuration file behaves the same way as the old defaults did.  If the
    settings in /etc/default/seedd have been customised, then on upgrade we

src/bbctl.cpp
//  This file is distributed as part of the bit-babbler package.
//  Copyright 2014 - 2018,  Ron <ron@debian.org>

#include "private_setup.h"

#include <bit-babbler/client-socket.h>
#include <bit-babbler/qa.h>

#include <bit-babbler/impl/log.h>

#include <unistd.h>
#include <getopt.h>


using BitB::Json;
using BitB::ClientSock;
using BitB::QA::Ent8;
using BitB::QA::Ent16;
using BitB::StrToU;
using BitB::StrToScaledU;
using BitB::StrToScaledUL;
using BitB::Error;
using BitB::Log;
using BitB::stringprintf;
using std::string;


static void usage()
{
    printf("Usage: bbctl [OPTION...]\n");
    printf("\n");
    printf("Query and control tool for BitBabbler hardware RNG devices\n");
    printf("\n");
    printf("Options:\n");
    printf("  -s, --scan                Scan for active devices\n");
    printf("  -i, --device-id=id        Act on only a single device\n");
    printf("  -b, --bin-freq            Report the 8-bit symbols sorted by frequency\n");
    printf("  -B, --bin-freq16          Report the 16-bit symbols sorted by frequency\n");
    printf("      --bin-count           Report the 8-bit symbols in symbol order\n");
    printf("      --bin-count16         Report the 16-bit symbols in symbol order\n");
    printf("      --first=n             Show only the first n bins\n");
    printf("      --last=n              Show only the last n bins\n");
    printf("  -r, --bit-runs            Report on runs of consecutive bits\n");
    printf("  -S, --stats               Report general QA statistics\n");
    printf("  -c, --control-socket=path The service socket to query\n");
    printf("  -V, --log-verbosity=n     Change the logging verbosity\n");
    printf("      --waitfor=dev:n:r:max Wait for a device to pass some number of bytes\n");
    printf("  -v, --verbose             Enable verbose output\n");
    printf("  -?, --help                Show this help message\n");
    printf("      --version             Print the program version\n");
    printf("\n");
    printf("Report bugs to support@bitbabbler.org\n");
    printf("\n");
}


struct WaitFor
{ //{{{

    typedef std::list< WaitFor >    List;

    std::string     deviceid;
    size_t          bytes;
    size_t          retry_ms;
    size_t          timeout_ms;

    WaitFor( const std::string &arg )
        : retry_ms( 1000 )
        , timeout_ms( 0 )
    { //{{{

        // Parse the options from a string of the form:
        // device:bytes:retry_ms:timeout_ms
        // Where device and bytes are mandatory.

        size_t  n = arg.find( ':' );

        if( n == string::npos )
            throw Error( _("No byte count given in --waitfor=%s"), arg.c_str() );

        deviceid = arg.substr( 0, n );
        ++n;

        try {
            size_t  n2 = arg.find( ':', n );

            if( n2 == string::npos )
            {
                bytes = StrToScaledUL( arg.substr(n), 1024 );
                return;
            }

            bytes   = StrToScaledUL( arg.substr(n, n2 - n), 1024 );
            n       = n2 + 1;
            n2      = arg.find( ':', n );

            if( n2 == string::npos )
            {
                retry_ms = StrToScaledUL( arg.substr(n) );
                goto done;
            }

            retry_ms    = StrToScaledUL( arg.substr(n, n2 - n) );
            n           = n2 + 1;
            timeout_ms  = StrToScaledUL( arg.substr(n) );

        done:
            if( retry_ms < 1 )
                throw Error( _("Retry time must be >= 1ms in --waitfor=%s"),
                                                                arg.c_str() );
        }
        catch( const std::exception &e )
        {
            throw Error( _("Invalid --waitfor argument '%s': %s"),
                                            arg.c_str(), e.what() );
        }

    } //}}}

}; //}}}


int main( int argc, char *argv[] )
{
  try {

    unsigned        opt_scan        = 0;
    unsigned        opt_bin_count   = 0;
    unsigned        opt_bin_freq    = 0;
    unsigned        opt_bit_runs    = 0;
    unsigned        opt_stats       = 0;
    unsigned        opt_first       = 65536;
    unsigned        opt_last        = 65536;
    unsigned        opt_log_level   = unsigned(-1);
    string          opt_deviceid;
    string          opt_controlsock = SEEDD_CONTROL_SOCKET;
    WaitFor::List   opt_wait;

    enum
    {
        BINCOUNT_OPT,
        BINCOUNT16_OPT,
        FIRST_OPT,
        LAST_OPT,
        WAITFOR_OPT,
        VERSION_OPT
    };

    struct option long_options[] =
    {
        { "scan",           no_argument,        NULL,      's' },
        { "device-id",      required_argument,  NULL,      'i' },
        { "bin-freq",       no_argument,        NULL,      'b' },
        { "bin-freq16",     no_argument,        NULL,      'B' },
        { "bin-count",      no_argument,        NULL,      BINCOUNT_OPT },
        { "bin-count16",    no_argument,        NULL,      BINCOUNT16_OPT },
        { "first",          required_argument,  NULL,      FIRST_OPT },
        { "last",           required_argument,  NULL,      LAST_OPT },
        { "bit-runs",       no_argument,        NULL,      'r' },
        { "stats",          no_argument,        NULL,      'S' },
        { "control-socket", required_argument,  NULL,      'c' },
        { "log-verbosity",  required_argument,  NULL,      'V' },
        { "waitfor",        required_argument,  NULL,      WAITFOR_OPT },
        { "verbose",        no_argument,        NULL,      'v' },
        { "help",           no_argument,        NULL,      '?' },
        { "version",        no_argument,        NULL,      VERSION_OPT },
        { 0, 0, 0, 0 }
    };

    int opt_index = 0;

    for(;;)
    { //{{{

        int c = getopt_long( argc, argv, ":si:c:bBrSV:v?",
                             long_options, &opt_index );
        if( c == -1 )
            break;

        switch(c)
        {
            case 's':
                opt_scan = 1;
                break;

            case 'i':
                opt_deviceid = optarg;
                break;

            case 'b':
                opt_bin_freq = 1;
                break;

            case 'B':
                opt_bin_freq = 16;
                break;

            case BINCOUNT_OPT:
                opt_bin_freq = 1;
                opt_bin_count = 1;
                break;

            case BINCOUNT16_OPT:
                opt_bin_freq = 16;
                opt_bin_count = 1;
                break;

            case FIRST_OPT:
                opt_first = StrToScaledU( optarg );
                if( opt_last == 65536 )
                    opt_last = 0;

                break;

            case LAST_OPT:
                opt_last = StrToScaledU( optarg );
                if( opt_first == 65536 )
                    opt_first = 0;
                break;

            case 'r':
                opt_bit_runs = 1;
                break;

            case 'S':
                opt_stats = 1;
                break;

            case 'c':
                opt_controlsock = optarg;
                break;

            case 'V':
                opt_log_level = StrToU( optarg, 10 );
                break;

            case WAITFOR_OPT:
                opt_wait.push_back( WaitFor(optarg) );
                break;

            case 'v':
                ++BitB::opt_verbose;
                break;

            case '?':
                if( optopt != '?' && optopt != 0 )
                {
                    fprintf(stderr, "%s: invalid option -- '%c', try --help\n",
                                                            argv[0], optopt);
                    return EXIT_FAILURE;
                }
                usage();
                return EXIT_SUCCESS;

            case ':':
                fprintf(stderr, "%s: missing argument for '%s', try --help\n",
                                                    argv[0], argv[optind - 1] );
                return EXIT_FAILURE;

            case VERSION_OPT:
                printf("bbctl " PACKAGE_VERSION "\n");
                return EXIT_SUCCESS;
        }

    } //}}}


    ClientSock      client( opt_controlsock );


    if( opt_log_level != unsigned(-1) )
    { //{{{

        client.SendRequest( stringprintf( "[\"SetLogVerbosity\",0,%u]",
                                                        opt_log_level ) );
        Json::Handle    json = client.Read();

        Log<4>("read reply: %s\n", json->JSONStr().c_str() );

        if( json[0]->String() == "SetLogVerbosity" )
        {
            printf( "Log verbosity is now %u\n", unsigned(json[2]) );

        } else {

            Log<0>( "unrecognised reply\n" );
        }

    } //}}}


    if( opt_scan )
    { //{{{

        client.SendRequest( "\"GetIDs\"" );

        Json::Handle    json = client.Read();

        Log<4>("read reply: %s\n", json->JSONStr().c_str() );

        if( json[0]->String() == "GetIDs" )
        {
            Json::Data::Handle  ids  = json[2];
            size_t              n    = ids->GetArraySize();

            printf( P_("Have %zu active device:\n",
                       "Have %zu active devices:\n", n), n );

            for( size_t i = 0; i < n; ++i )
                printf( _("  Device ID: %s\n"), ids[i]->String().c_str() );

        } else {

            Log<0>( "unrecognised reply\n" );
        }

    } //}}}


    while( ! opt_wait.empty() )
    { //{{{

        const WaitFor   &w = opt_wait.front();
        size_t          elapsed;

        if( w.timeout_ms )
            Log<1>( _("Waiting up to %zu ms for %zu good bytes from %s\n"),
                                w.timeout_ms, w.bytes, w.deviceid.c_str() );
        else
            Log<1>( _("Waiting for %zu good bytes from %s\n"),
                                              w.bytes, w.deviceid.c_str() );

        for( elapsed = 0; w.timeout_ms == 0 || elapsed < w.timeout_ms;
                                                        elapsed += w.retry_ms )
        {
            client.SendRequest( "[\"ReportStats\",1,\"" + w.deviceid + "\"]" );

            Json::Handle    json = client.Read();

            Log<4>("read reply: %s\n", json->JSONStr().c_str() );

            if( json[0]->String() == "ReportStats" )
            {
                Json::Data::Handle  stats = json[2]->Get( w.deviceid );

                if( ! stats )
                    throw Error( _("No statistics available for device '%s'"),
                                                         w.deviceid.c_str() );

                unsigned long long  passed = stats["QA"]["BytesPassed"]->
                                                    As<unsigned long long>();
                if( passed >= w.bytes )
                {
                    Log<1>( _("Have %llu good bytes from %s in %zums\n"),
                                    passed, w.deviceid.c_str(), elapsed );
                    goto done;
                }

                Log<3>( _("Have %llu good bytes from %s in %zums (waiting for %zu)\n"),
                                        passed, w.deviceid.c_str(), elapsed, w.bytes );
            } else {

                // Possibly we should throw here too, but since this should
                // never happen, assume it's a glitch and just try again.
                Log<0>( "Unrecognised reply to ReportStats request\n" );
            }

            usleep( useconds_t( w.retry_ms * 1000 ) );
        }

        throw Error( _("Timeout after %zums waiting for %zu bytes from %s\n"),
                                    elapsed, w.bytes, w.deviceid.c_str() );

    done:
        opt_wait.pop_front();

    } //}}}


    if( opt_bin_freq )
    { //{{{

        if( opt_deviceid.empty() )
            client.SendRequest( "\"GetRawData\"" );
        else
            client.SendRequest( "[\"GetRawData\",1,\"" + opt_deviceid + "\"]" );

        Json::Handle    json = client.Read();

        Log<4>("read reply: %s\n", json->JSONStr().c_str() );

        if( json[0]->String() == "GetRawData" )
        {
            Json::Data::Handle  data = json[2];
            Json::MemberList    sources;

            data->GetMembers( sources );

            for( Json::MemberList::iterator si = sources.begin(),
                                            se = sources.end(); si != se; ++si )
            {
                Json::Data::Handle  ent8  = data[*si]->Get("Ent8");
                Json::Data::Handle  ent16 = data[*si]->Get("Ent16");

                if( ! ent8 )
                {
                    printf( "\nsource: %s has no 8-bit data (yet)\n", si->c_str() );

                } else {

                    Ent8::Data  e8_short( ent8["Short"] );
                    Ent8::Data  e8_long( ent8["Long"] );

                    if( opt_bin_count )
                    {
                        printf( "\nsource: %s\n%s\n",
                                si->c_str(), e8_short.ReportBins( opt_first, opt_last ).c_str() );
                        printf( "\nsource: %s\n%s\n",
                                si->c_str(), e8_long.ReportBins( opt_first, opt_last ).c_str() );

                    } else {

                        printf( "\nsource: %s\n%s\n",
                                si->c_str(), e8_short.ReportBinsByFreq( opt_first, opt_last ).c_str() );
                        printf( "\nsource: %s\n%s\n",
                                si->c_str(), e8_long.ReportBinsByFreq( opt_first, opt_last ).c_str() );
                    }
                }

                if( opt_bin_freq == 16 )
                {
                    if( ! ent16 )
                    {
                        printf( "\nsource: %s has no 16-bit data (yet)\n", si->c_str() );

                    } else {

                        Ent16::Data  e16_short( ent16["Short"] );
                        Ent16::Data  e16_long( ent16["Long"] );

                        if( opt_bin_count )
                        {
                            printf( "\nsource: %s\n%s\n",
                                    si->c_str(), e16_short.ReportBins( opt_first, opt_last ).c_str() );
                            printf( "\nsource: %s\n%s\n",
                                    si->c_str(), e16_long.ReportBins( opt_first, opt_last ).c_str() );

                        } else {

                            printf( "\nsource: %s\n%s\n",
                                    si->c_str(), e16_short.ReportBinsByFreq( opt_first, opt_last ).c_str() );
                            printf( "\nsource: %s\n%s\n",
                                    si->c_str(), e16_long.ReportBinsByFreq( opt_first, opt_last ).c_str() );
                        }
                    }
                }
            }

        } else {

            Log<0>( "unrecognised reply\n" );
        }

    } //}}}


    if( opt_bit_runs )
    { //{{{

        using BitB::QA::BitRuns;

        if( opt_deviceid.empty() )
            client.SendRequest( "\"ReportStats\"" );
        else
            client.SendRequest( "[\"ReportStats\",1,\"" + opt_deviceid + "\"]" );

        Json::Handle    json = client.Read();

        Log<4>("read reply: %s\n", json->JSONStr().c_str() );

        if( json[0]->String() == "ReportStats" )
        {
            Json::Data::Handle  stats = json[2];
            Json::MemberList    sources;

            stats->GetMembers( sources );

            for( Json::MemberList::iterator si = sources.begin(),
                                            se = sources.end(); si != se; ++si )
            {
                BitRuns::Result     bitruns( stats[*si]["BitRuns"] );

                printf( "\nsource: %s\n%s\n", si->c_str(), bitruns.Report().c_str() );
            }

        } else {

            Log<0>( "unrecognised reply\n" );
        }

    } //}}}


    if( opt_stats )
    { //{{{

        if( opt_deviceid.empty() )
            client.SendRequest( "\"ReportStats\"" );
        else
            client.SendRequest( "[\"ReportStats\",1,\"" + opt_deviceid + "\"]" );

        Json::Handle    json = client.Read();

        Log<4>("read reply: %s\n", json->JSONStr().c_str() );

        if( json[0]->String() == "ReportStats" )
        {
            Json::Data::Handle  stats = json[2];
            Json::MemberList    sources;

            stats->GetMembers( sources );

            for( Json::MemberList::iterator si = sources.begin(),
                                            se = sources.end(); si != se; ++si )
            {
                unsigned long long  analysed = stats[*si]["QA"]["BytesAnalysed"]->
                                                            As<unsigned long long>();
                unsigned long long  passed   = stats[*si]["QA"]["BytesPassed"]->
                                                            As<unsigned long long>();
                BitB::QA::FIPS      fips( stats[*si]["FIPS"] );

                printf( "\nsource: %s\n", si->c_str() );

                printf( "Octets analysed %llu, passed %llu, (not passed %llu)\n",
                                            analysed, passed, analysed - passed );

                printf( "FIPS %s\n", fips.ReportFailRates().c_str() );
                printf( "FIPS %s\n", fips.ReportPassRuns().c_str() );

                Json::Data::Handle  ent8  = stats[*si]->Get("Ent8");
                Json::Data::Handle  ent16 = stats[*si]->Get("Ent16");

                if( ! ent8 )
                {
                    printf( "Ent8: no results (yet)\n" );

                } else {

                    Ent8::Data  e8_short( Ent8::Results_Only, ent8["Short"] );
                    Ent8::Data  e8_long( Ent8::Results_Only, ent8["Long"] );

                    printf( "Ent8 short %s\n", e8_short.ReportResults().c_str() );
                    printf( "Ent8 long %s\n", e8_long.ReportResults().c_str() );
                }

                if( ! ent16 )
                {
                    printf( "Ent16: no results (yet)\n" );

                } else {

                    Ent16::Data  e16_short( Ent16::Results_Only, ent16["Short"] );
                    Ent16::Data  e16_long( Ent16::Results_Only, ent16["Long"] );

                    printf( "Ent16 short %s\n", e16_short.ReportResults().c_str() );
                    printf( "Ent16 long %s\n", e16_long.ReportResults().c_str() );
                }
            }

        } else {

            Log<0>( "unrecognised reply\n" );
        }

    } //}}}


    return EXIT_SUCCESS;
  }
  BB_CATCH_ALL( 0, _("bbctl fatal exception") )

  return EXIT_FAILURE;
}

// vi:sts=4:sw=4:et:foldmethod=marker
